### PR TITLE
example a11y affordances

### PR DESCRIPTION
**Pull request by declan-warn**

### CODE REVIEW

File: pieces/scenario/react-beautiful-dnd/card.tsx [hunk 1]
```diff
@@ -67,6 +68,12 @@ function CardText({ state }: { state: DraggableState }) {
   );
 }
 
+const controlStyles = css({
+  position: 'absolute',
+  top: 8,
```
Comment (alexreardon) [inline pieces/scenario/react-beautiful-dnd/card.tsx]: can we move these numbers (and others) to be based on `var(--grid)`?

File: pieces/shared/menu-button/container.tsx [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+      case 'Home':
+        menuRef.current?.focusFirst();
+        break;
+
+      case 'End':
+        menuRef.current?.focusLast();
+        break;
+    }
+
+    if (containerRef.current && /[a-z]/i.test(event.key)) {
```
Comment (alexreardon) [inline pieces/shared/menu-button/container.tsx]: Q: can you explain what this is achieving?

File: pieces/shared/menu-button/trigger.tsx [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    <Button
+      role="button"
+      aria-haspopup="menu"
+      aria-expanded={isOpen}
+      aria-label={label}
+      onClick={onClick}
+      onKeyDown={onKeyDown}
+      ref={ref}
+    >
+      <img {...moreIcon} alt="" draggable={false} />
```
Comment (alexreardon) [inline pieces/shared/menu-button/trigger.tsx:83]: should we add a `role="presentation"` to the image?

**Review by alexreardon — CHANGES_REQUESTED** [review]

**Comment by alexreardon** [comment]
Can the generic actions go before the movement actions?

Screen Shot 2022-09-28 at 1 28 37 pm

**Comment by alexreardon** [comment]
Can you please add some generic actions to the `...` buttons on cards as well?

File: pieces/shared/menu-button/container.tsx [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+
+import Trigger from './trigger';
+import Menu, { MenuHandle } from './menu';
+
+const containerStyles = css({
+  display: 'flex',
+  position: 'relative',
+});
+
+const useInitialRender = () => {
```
Comment (alexreardon) [inline pieces/shared/menu-button/container.tsx]: why is this needed?

File: pieces/shared/menu-button/container.tsx [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+  const [initialFocus, setInitialFocus] = useState<'first' | 'last'>('first');
+
+  const menuRef = useRef<MenuHandle>(null);
+
+  const triggerRef = useRef<HTMLButtonElement>(null);
+
+  const shouldResetFocusRef = useRef(false);
+
+  useEffect(() => {
+    if (!isOpen && shouldResetFocusRef.current) {
```
Comment (alexreardon) [inline pieces/shared/menu-button/container.tsx]: why is this needed?

File: pieces/shared/menu-button/container.tsx [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+      shouldResetFocusRef.current = false;
+      window.setTimeout(() => {
+        triggerRef.current?.focus();
+      }, 0);
+    }
+  }, [isOpen]);
+
+  const onKeyDown: KeyboardEventHandler<HTMLButtonElement> = useCallback((event) => {
+    switch (event.key) {
```
Comment (alexreardon) [inline pieces/shared/menu-button/container.tsx]: personal preference: can we use `if` statments here?

Comment (alexreardon) [reply pieces/shared/menu-button/container.tsx]: Similar to what you do in the following `useEffect`

File: pieces/shared/menu-button/container.tsx [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+      }
+    }
+  }, []);
+
+  useEffect(() => {
+    if (!isOpen) {
+      return;
+    }
+
+    window.addEventListener('click', toggleIsOpen);
```
Comment (alexreardon) [inline pieces/shared/menu-button/container.tsx]: For safety, can this please use `bind-event-listener`?

File: pieces/shared/menu-button/container.tsx [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+
+    return () => {
+      window.removeEventListener('click', toggleIsOpen);
+    };
+  }, [isOpen]);
+
+  const onClick: MouseEventHandler = useCallback((event) => {
+    toggleIsOpen();
+    event.stopPropagation();
```
Comment (alexreardon) [inline pieces/shared/menu-button/container.tsx]: why is the event stopped rather than prevented?

File: pieces/shared/menu-button/menu-item.tsx [hunk 9]
```diff
@@ -0,0 +1,47 @@
+import { FocusEventHandler, ReactNode, useCallback, useState } from 'react';
+
+import { css } from '@emotion/react';
+
+import { token } from '@atlaskit/tokens';
+
+const menuItemStyles = css({
+  display: 'flex',
+  padding: 8,
```
Comment (alexreardon) [inline pieces/shared/menu-button/menu-item.tsx]: please move to `--grid`

File: pieces/shared/menu-button/menu.tsx [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+  event.stopPropagation();
+};
+
+const Menu = forwardRef<MenuHandle, MenuProps>(function Menu(
+  { children, onKeyDown: onKeydownProp, onClose },
+  handleRef,
+) {
+  const internalRef = useRef<HTMLUListElement>(null);
+
+  useImperativeHandle(
```
Comment (alexreardon) [inline pieces/shared/menu-button/menu.tsx]: I generally find imperative handles to be a bit mind bendy. Can you please switch this just to be helper functions?

```ts
function isHTMLElement(el: Node | null): el is HTMLElement {
  return el instanceof HTMLElement;
}

function focusFirstChild(el: Element | null) {
  if (!el) {
    return;
  }

  if (isHTMLElement(el.firstChild)) {
    el.firstChild.focus();
  }
}

function focusLastChild(el: Element | null) {
  if (!el) {
    return;
  }

  if (isHTMLElement(el.lastElementChild)) {
    el.lastElementChild.focus();
  }
}

function focusNextChild(el: Element | null) {
  if (!el) {
    return;
  }
  const focused = el.querySelector(":focus");
  // error case
  if (!focused) {
    focusFirstChild(el);
    return;
  }
  const next = el.nextElementSibling;
  if (!isHTMLElement(next)) {
    focusFirstChild(el);
    return;
  }
}

```

File: pieces/shared/menu-button/menu.tsx [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+          onClose({ shouldResetFocus: true });
+          break;
+      }
+    },
+    [onKeydownProp, onClose],
+  );
+
+  const onClick: MouseEventHandler = useCallback(
+    (event) => {
+      event.stopPropagation();
```
Comment (alexreardon) [inline pieces/shared/menu-button/menu.tsx]: why is the event stopped?

File: pieces/shared/menu-button/container.tsx [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+
+    if (initialFocus === 'last') {
+      menuRef.current?.focusLast();
+    }
+  }, [initialFocus, isOpen]);
+
+  const containerRef = useRef<HTMLSpanElement>(null);
+
+  const onMenuKeyDown: KeyboardEventHandler<HTMLButtonElement> = useCallback((event) => {
+    switch (event.key) {
```
Comment (alexreardon) [inline pieces/shared/menu-button/container.tsx]: Preference: please also move from switch to if

**Review by alexreardon — CHANGES_REQUESTED** [review]

File: pieces/shared/menu-button/container.tsx (hunk 2 above)
Comment (declan-warn) [reply pieces/shared/menu-button/container.tsx]: added a comment in the code, but this will select the next menu item which starts with the letter pressed

File: pieces/shared/menu-button/trigger.tsx (hunk 3 above)
Comment (declan-warn) [reply pieces/shared/menu-button/trigger.tsx:83]: I don't think that is needed if we use `alt=""` (https://www.w3.org/WAI/tutorials/images/decorative/)

File: pieces/shared/menu-button/container.tsx (hunk 4 above)
Comment (declan-warn) [reply pieces/shared/menu-button/container.tsx]: it's not, I removed it :)

File: pieces/shared/menu-button/container.tsx (hunk 5 above)
Comment (declan-warn) [reply pieces/shared/menu-button/container.tsx]: hopefully the changed code should make this clearer, but it's to handle moving focus back to the trigger when the menu closes

File: pieces/shared/menu-button/container.tsx (hunk 8 above)
Comment (declan-warn) [reply pieces/shared/menu-button/container.tsx]: I don't remember exactly why I was doing this, but have removed it now

File: pieces/shared/menu-button/menu.tsx (hunk 11 above)
Comment (declan-warn) [reply pieces/shared/menu-button/menu.tsx]: the original intention was so that we could ignore clicks from inside the menu, but I've refactored it to just do a `.contains` check instead.

**Comment by alexreardon** [comment]
Looking good!

A few UX things:

- Can you please add 'move left' + 'move right' (or 'Move to Column B') to cards?
- Can you please add the movement items (move down etc) to columns and cards for react-dnd and deferred-atlaskit?

File: pieces/shared/menu-button/container.tsx [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+
+function reducer(_: MenuButtonState, action: MenuButtonAction): MenuButtonState {
+  if (action.type === 'open') {
+    return {
+      isOpen: true,
+      initialFocus: action.initialFocus,
+    };
+  }
+
+  if (action.type === 'close') {
```
Comment (alexreardon) [inline pieces/shared/menu-button/container.tsx]: You don't need this guard. Typescript whill know that at this point the only action.type available is 'close'. 

So you can remove the guard and the throw below as well

File: pieces/shared/menu-button/focus.tsx [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+  const active = el?.querySelector('[tabindex="0"]');
+  if (active?.previousElementSibling) {
+    focus(active.previousElementSibling);
+  } else {
+    focusLastItem(el);
+  }
+}
+
+/**
+ * When pressing a letter (A-Z or a-z) then:
```
Comment (alexreardon) [inline pieces/shared/menu-button/focus.tsx:38]: Is this _needed_ for accessibility? If not, I think we should drop this

File: pieces/shared/menu-button/trigger.tsx [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+    isOpen: boolean;
+    shouldResetFocus: boolean;
+  },
+) {
+  useEffect(() => {
+    if (!isOpen && shouldResetFocus) {
+      // Running in an 'immediate' timeout so that
+      // when the menu has closed because of an enter key press,
+      // it won't trigger a click event on the trigger as well.
+      setTimeout(() => {
```
Comment (alexreardon) [inline pieces/shared/menu-button/trigger.tsx:38]: I wonder if there is a way to do this without relying on a setTimeout. All good to leave this though

**Review by alexreardon — CHANGES_REQUESTED** [review]

File: pieces/shared/menu-button/focus.tsx (hunk 14 above)
Comment (declan-warn) [reply pieces/shared/menu-button/focus.tsx:38]: I cannot find exactly where in the W3 spec this is required, but we have it in our Accessibility Standards and also https://www.w3.org/WAI/ARIA/apg/example-index/menu-button/menu-button-actions.html.

File: pieces/shared/menu-button/trigger.tsx (hunk 15 above)
Comment (declan-warn) [reply pieces/shared/menu-button/trigger.tsx:38]: I think `event.preventDefault()` on the `keydown` event works as well.

File: pieces/shared/menu-button/container.tsx (hunk 13 above)
Comment (declan-warn) [reply pieces/shared/menu-button/container.tsx]: fixed :D

File: pieces/scenario/atlaskit-drag-and-drop/card.tsx [hunk 16]
```diff
@@ -134,6 +141,19 @@ export const Card = memo(function Card({ item }: { item: Item }) {
       <DragIcon state={state} />
       <CardText state={state} />
       <DropIndicator edge={closestEdge} gap={'var(--card-gap)'} />
+      <div css={controlStyles}>
+        <MenuButton label={`controls for card ${itemId}`}>
+          <MenuItem>Edit</MenuItem>
+          <MenuItem>Share</MenuItem>
+          <MenuItem>Move up</MenuItem>
+          <MenuItem>Move down</MenuItem>
+          {['A', 'B', 'C']
```
Comment (alexreardon) [inline pieces/scenario/atlaskit-drag-and-drop/card.tsx]: Nice! Not overly complicated and gets the job done 👍 

A thought: could we pass all column ids as an array to each card? That way we don't need to hard code them

What do you think?

**Review by alexreardon — APPROVED** [review]
This is looking good. I added a minor thing for you to consider. Once the build is green I am happy for this to merge

File: pieces/scenario/atlaskit-drag-and-drop/card.tsx (hunk 16 above)
Comment (declan-warn) [reply pieces/scenario/atlaskit-drag-and-drop/card.tsx]: I agree not hard coding is better. I originally did it as a quick solution, but have revised it now to pass them down :)

**Comment by declan-warn** [comment]
@alexreardon I think there is a config issue causing the validate workflow to fail. Any thoughts?

```
$ next lint
/bin/sh: 1: next: not found
```

Do we need to run `yarn` explicitly?

File: pieces/scenario/deferred-atlaskit-drag-and-drop/card.tsx [hunk 17]
```diff
@@ -99,6 +115,19 @@ export const Card = memo(function Card({ item }: { item: Item }) {
       <Suspense>
         <LazyDropIndicator edge={closestEdge} gap={'var(--card-gap)'} />
       </Suspense>
+      <div css={controlStyles}>
+        <MenuButton label={`controls for card ${itemId}`}>
+          <MenuItem>Edit</MenuItem>
+          <MenuItem>Share</MenuItem>
+          <MenuItem>Move up</MenuItem>
+          <MenuItem>Move down</MenuItem>
+          {orderedColumnIds
```
Comment (alexreardon) [inline pieces/scenario/deferred-atlaskit-drag-and-drop/card.tsx:124]: I am now thinking that maybe that using `orderedColumnIds` might not be ideal as it would mean that all cards in a column would need to re-render when a column is reordered

Comment (declan-warn) [reply pieces/scenario/deferred-atlaskit-drag-and-drop/card.tsx:124]: I've added a comment about this, as discussed :)
